File: tutorials/module1/scala/utilities/DomainObjects.scala
package module1.scala.utilities

import java.time.Instant
import java.time.format.DateTimeParseException
import scala.collection.mutable
import module1.scala.utilities.WarcRecord.getCommonFields

/**
 * Domain objects for the web corpus
 *
 * @author Phil, https://github.com/g1thubhub
 */

case class WarcRecord(warcType: String, dateS: Long, recordID: String, contentLength: Int, contentType: String, infoID: String, concurrentTo: String, ip: String, targetURI: String, payloadDigest: String, blockDigest: String, payloadType: String, htmlContentType: String, language: Option[String], htmlLength: Int, htmlSource: String)

object WarcRecord {

  def getCommonFields(metaPairs: mutable.Map[String, String]): (String, String, String, String, String, Long, Int) = {
    val warcType = metaPairs.getOrElse("WARC-Type", "")
    val targetURI = metaPairs.getOrElse("WARC-Target-URI", "")
    val recordID = metaPairs.getOrElse("WARC-Record-ID", "")
    val contentType = metaPairs.getOrElse("Content-Type", "")
    val blockDigest = metaPairs.getOrElse("WARC-Block-Digest", "")
    var dateS = -1L
    var contentLength = -1
    try {
      contentLength = metaPairs.getOrElse("Content-Length", "-1").toInt
      dateS = Instant.parse(metaPairs.getOrElse("WARC-Date", "+1000000000-12-31T23:59:59.999999999Z")).getEpochSecond
    }
    catch {
      case _: NumberFormatException => System.err.println(s"Malformed contentLength field for record $recordID")
      case _: DateTimeParseException => System.err.println(s"Malformed date field for record $recordID")
      case e: Exception => e.printStackTrace()
    }
    (warcType, targetURI, recordID, contentType, blockDigest, dateS, contentLength)
  }

  def apply(metaPairs: mutable.Map[String, String], responseMeta: (String, Option[String], Int), sourceHtml: String): WarcRecord = {
    val (warcType, targetURI, recordID, contentType, blockDigest, dateS, contentLength) = getCommonFields(metaPairs)
    val payloadDigest = metaPairs.getOrElse("WARC-Payload-Digest", "")
    val infoID = metaPairs.getOrElse("WARC-Warcinfo-ID", "")
    val concurrentTo = metaPairs.getOrElse("WARC-Concurrent-To", "")
    val ip = metaPairs.getOrElse("WARC-IP-Address", "")
    val payloadType = metaPairs.getOrElse("WARC-Identified-Payload-Type", "")

    WarcRecord(warcType, dateS, recordID, contentLength, contentType, infoID, concurrentTo, ip, targetURI, payloadDigest, blockDigest, payloadType, responseMeta._1, responseMeta._2, responseMeta._3, sourceHtml)
  }
}

case class WetRecord(warcType: String, targetURI: String, dateS: Long, recordID: String, refersTo: String, blockDigest: String, contentType: String, contentLength: Int, plainText: String)

object WetRecord {
  def apply(metaPairs: mutable.Map[String, String], pageContent: String): WetRecord = {
    val (warcType, targetURI, recordID, contentType, blockDigest, dateS, contentLength) = getCommonFields(metaPairs)
    val refersTo = metaPairs.getOrElse("WARC-Refers-To", "")
    WetRecord(warcType, targetURI, dateS, recordID, refersTo, blockDigest, contentType, contentLength, pageContent)
  }

  import java.util.UUID.randomUUID
  import scala.util.Random
  def createDummy() = WetRecord(randomUUID().toString, randomUUID().toString, System.currentTimeMillis(), randomUUID().toString, randomUUID().toString, randomUUID().toString, randomUUID().toString, Random.nextInt(), randomUUID().toString)

}

File: tutorials/module1/python/utilities/domain_objects.py
from datetime import datetime
from typing import Tuple, DefaultDict
from uuid import uuid4
from random import randrange
from time import strftime

# Domain objects for the web corpus
# @author Phil, https://github.com/g1thubhub

default_date = "1970-01-01T00:00:00Z"


def to_epochs(time_string):
    parsed_time = datetime.strptime(time_string, "%Y-%m-%dT%H:%M:%S%z")
    return int(parsed_time.timestamp())


def get_common_field(meta_pairs: DefaultDict[str, str]):
    warc_type = meta_pairs.get("WARC-Type", "")
    target_uri = meta_pairs.get("WARC-Target-URI", "")
    record_id = meta_pairs.get("WARC-Record-ID", "")
    content_type = meta_pairs.get("Content-Type", "")
    block_digest = meta_pairs.get("WARC-Block-Digest", "")
    date = meta_pairs.get("WARC-Date", default_date)
    date_s = to_epochs(date)
    content_length = int(meta_pairs.get("Content-Length", "-1"))
    return warc_type, target_uri, record_id, content_type, block_digest, date_s, content_length


class WarcRecord:
    def __init__(self, meta_pairs: DefaultDict[str, str], response_meta: Tuple[str, str, int], source_html: str):
        warc_type, target_uri, record_id, content_type, block_digest, date_s, content_length = get_common_field(
            meta_pairs)
        self.warc_type = warc_type
        self.target_uri = target_uri
        self.record_id = record_id
        self.content_type = content_type
        self.block_digest = block_digest
        self.date_s = date_s
        self.content_length = content_length

        self.info_id = meta_pairs.get("WARC-Warcinfo-ID", "")
        self.concurrent_to = meta_pairs.get("WARC-Concurrent-To", "")
        self.ip = meta_pairs.get("WARC-IP-Address", "")
        self.payload_digest = meta_pairs.get("WARC-Payload-Digest", "")
        self.payload_type = meta_pairs.get("WARC-Identified-Payload-Type", "")
        self.html_content_type = response_meta[0]
        self.language = response_meta[1]
        self.html_length = response_meta[2]
        self.html_source = source_html


class WetRecord:
    def __init__(self, meta_pairs: DefaultDict[str, str], source_html: str):
        warc_type, target_uri, record_id, content_type, block_digest, date_s, content_length = get_common_field(
            meta_pairs)
        self.warc_type = warc_type
        self.target_uri = target_uri
        self.record_id = record_id
        self.content_type = content_type
        self.block_digest = block_digest
        self.date_s = date_s
        self.content_length = content_length
        self.plain_text = source_html
        self.refers_to = meta_pairs.get("WARC-Refers-To", "")

    @staticmethod
    def create_dummy():
        dict = {"WARC-Type": str(uuid4()), "WARC-Target-URI": str(uuid4()), "WARC-Record-ID": str(uuid4()),
                "Content-Type": str(uuid4()),
                "WARC-Block-Digest": str(uuid4()), "WARC-Date": strftime("%Y-%m-%dT%H:%M:%S%z"),
                "Content-Length": str(randrange(1000000))}
        return WetRecord(dict, str(uuid4()))
